File: packages/sdk/src/resources/sessions.ts
import type { TheatricalHTTPClient } from '../http/client';
import type { Session, SessionFilter, SessionListResponse, SeatAvailability } from '../types/session';
import { sessionSchema, sessionListResponseSchema, seatAvailabilitySchema } from '../types/session';
import type { PaginatedResponse, PaginationParams } from '../types/pagination';

/** Default page size for auto-paginated iteration */
const DEFAULT_PAGE_SIZE = 50;

/**
 * Sessions resource — showtimes, availability, and seat maps.
 *
 * @example
 * ```typescript
 * const sessions = await client.sessions.list({
 *   siteId: 'roxy-wellington',
 *   date: '2026-04-08',
 * });
 *
 * const seats = await client.sessions.availability(sessions[0].id);
 * ```
 *
 * @example Auto-paginate all sessions
 * ```typescript
 * for await (const session of client.sessions.listAll({ siteId: 'roxy-wellington' })) {
 *   console.log(session.filmTitle, session.startTime);
 * }
 * ```
 */
export class SessionsResource {
  constructor(private readonly http: TheatricalHTTPClient) {}

  /**
   * List sessions (showtimes) with optional filters.
   * Response is validated at runtime using Zod — malformed API responses throw a parse error.
   * @see https://developer.vista.co/digital-platform/sessions/
   */
  async list(filters?: SessionFilter): Promise<SessionListResponse> {
    const raw = await this.http.get<unknown>('/ocapi/v1/sessions', {
      params: filters as Record<string, string | number | boolean | undefined>,
    });
    return sessionListResponseSchema.parse(raw);
  }

  /**
   * List sessions with explicit pagination control.
   *
   * Returns a normalised `PaginatedResponse<Session>` envelope that includes
   * the pagination strategy and cursor/offset for the next page.
   *
   * @param filters - Session filters (site, film, date range, etc.)
   * @param pagination - Page size and cursor/offset position
   * @returns A single page of sessions with pagination metadata
   *
   * @example
   * ```typescript
   * const page = await client.sessions.listPaginated(
   *   { siteId: 'roxy-wellington' },
   *   { limit: 25 },
   * );
   * console.log(page.data.length, page.hasMore);
   * ```
   */
  async listPaginated(
    filters?: SessionFilter,
    pagination?: PaginationParams,
  ): Promise<PaginatedResponse<Session>> {
    const limit = pagination?.limit ?? DEFAULT_PAGE_SIZE;

    const useCursor = pagination?.cursor !== undefined;

    const response = await this.list({
      ...filters,
      limit,
      ...(useCursor
        ? { cursor: pagination!.cursor }
        : { offset: pagination?.offset ?? 0 }),
    });

    return {
      data: response.sessions,
      total: response.total,
      hasMore: response.hasMore,
      nextCursor: response.nextCursor,
      nextOffset: response.nextOffset,
      strategy: useCursor ? 'cursor' : 'offset',
    };
  }

  /**
   * Async generator that automatically paginates through all matching sessions.
   *
   * Yields individual `Session` objects, fetching the next page transparently
   * when the current page is exhausted. Uses offset-based pagination internally
   * since Vista's session endpoints support numeric offsets.
   *
   * @param filters - Session filters (site, film, date range, etc.)
   * @param pageSize - Number of sessions to fetch per page (default: 50, max: 500)
   *
   * @example
   * ```typescript
   * const allSessions: Session[] = [];
   * for await (const session of client.sessions.listAll({ siteId: 'roxy-wellington' })) {
   *   allSessions.push(session);
   * }
   * ```
   *
   * @example Collect with early termination
   * ```typescript
   * for await (const session of client.sessions.listAll({ date: '2026-04-10' })) {
   *   if (session.isSoldOut) continue;
   *   console.log(`${session.filmTitle} — ${session.screenName} @ ${session.startTime}`);
   *   if (!session.isBookable) break; // stop on first unbookable
   * }
   * ```
   */
  async *listAll(
    filters?: SessionFilter,
    pageSize: number = DEFAULT_PAGE_SIZE,
  ): AsyncGenerator<Session, void, undefined> {
    let offset = 0;
    let hasMore = true;

    while (hasMore) {
      const response = await this.list({
        ...filters,
        limit: pageSize,
        offset,
      });

      for (const session of response.sessions) {
        yield session;
      }

      hasMore = response.hasMore;
      offset = response.nextOffset ?? offset + response.sessions.length;
    }
  }

  /**
   * Get a single session by ID.
   * Response is validated at runtime using Zod.
   */
  async get(sessionId: string): Promise<Session> {
    const raw = await this.http.get<unknown>(`/ocapi/v1/sessions/${sessionId}`);
    return sessionSchema.parse(raw);
  }

  /**
   * Get seat availability for a session.
   * Returns the complete auditorium seat map with status for each seat.
   * Response is validated at runtime using Zod.
   */
  async availability(sessionId: string): Promise<SeatAvailability> {
    const raw = await this.http.get<unknown>(`/ocapi/v1/sessions/${sessionId}/seat-plan`);
    return seatAvailabilitySchema.parse(raw);
  }
}

File: packages/sdk-py/theatrical/resources/sessions.py
"""Sessions resource — showtimes, availability, and seat maps."""

from __future__ import annotations

from collections.abc import AsyncIterator
from typing import TYPE_CHECKING, Optional

from pydantic import TypeAdapter

from theatrical.types.pagination import PaginatedResponse, PaginationParams, PaginationStrategy
from theatrical.types.session import (
    SeatAvailability,
    Session,
    SessionFilter,
    SessionListResponse,
)

if TYPE_CHECKING:
    from theatrical.http.client import TheatricalHttpProtocol

DEFAULT_PAGE_SIZE = 50

_session_list_response_adapter = TypeAdapter(SessionListResponse)
_session_adapter = TypeAdapter(Session)
_seat_availability_adapter = TypeAdapter(SeatAvailability)


class SessionsResource:
    def __init__(self, http: TheatricalHttpProtocol) -> None:
        self._http = http

    async def list(self, filters: Optional[SessionFilter] = None) -> SessionListResponse:
        params = _build_filter_params(filters) if filters else None
        raw = await self._http.get("/ocapi/v1/sessions", params=params)
        return _session_list_response_adapter.validate_python(raw)

    async def list_paginated(
        self,
        filters: Optional[SessionFilter] = None,
        pagination: Optional[PaginationParams] = None,
    ) -> PaginatedResponse[Session]:
        limit = pagination.limit if pagination and pagination.limit else DEFAULT_PAGE_SIZE
        use_cursor = pagination is not None and pagination.cursor is not None

        merged = SessionFilter(**(filters.model_dump(exclude_none=True) if filters else {}))
        merged.limit = limit
        if use_cursor:
            merged.cursor = pagination.cursor  # type: ignore[union-attr]
        else:
            merged.offset = (pagination.offset if pagination and pagination.offset is not None else 0)

        response = await self.list(merged)
        return PaginatedResponse[Session](
            data=response.sessions,
            total=response.total,
            has_more=response.has_more,
            next_cursor=response.next_cursor,
            next_offset=response.next_offset,
            strategy=PaginationStrategy.CURSOR if use_cursor else PaginationStrategy.OFFSET,
        )

    async def list_all(
        self,
        filters: Optional[SessionFilter] = None,
        page_size: int = DEFAULT_PAGE_SIZE,
    ) -> AsyncIterator[Session]:
        offset = 0
        has_more = True

        while has_more:
            merged = SessionFilter(**(filters.model_dump(exclude_none=True) if filters else {}))
            merged.limit = page_size
            merged.offset = offset

            response = await self.list(merged)
            for session in response.sessions:
                yield session

            has_more = response.has_more
            offset = response.next_offset if response.next_offset is not None else offset + len(response.sessions)

    async def get(self, session_id: str) -> Session:
        raw = await self._http.get(f"/ocapi/v1/sessions/{session_id}")
        return _session_adapter.validate_python(raw)

    async def availability(self, session_id: str) -> SeatAvailability:
        raw = await self._http.get(f"/ocapi/v1/sessions/{session_id}/seat-plan")
        return _seat_availability_adapter.validate_python(raw)


def _build_filter_params(filters: SessionFilter) -> dict[str, str]:
    params: dict[str, str] = {}
    d = filters.model_dump(exclude_none=True)
    for key, value in d.items():
        params[key] = str(value) if not isinstance(value, str) else value
    return params
